### PR TITLE
[android] Update to LTS NDK 27

### PR DESCRIPTION
**Pull request by finagolfin**
Add a new `bits/` header to the Android overlay, include runtime libraries that are
auto-extracted and listed many times to the list of libraries to be de-duplicated,
enable a C++ interop test that's working again, and update the doc with new
libraries that need to be available to run a simple executable.

@compnerd and @hyp, I [switched over to the latest NDK 27 with this patch a couple months ago](https://github.com/finagolfin/swift-android-sdk/commit/606ccefea792a7791cfbd49014f2e89d4347f2c7#diff-964afbea403833de16a6303ebb0c7f740abbf37e652520bb815db206aa268de1R10), no problem, plus it still worked fine with NDK 26 too. Let me know if you're using NDK 27 yet and please review.

@etcwilde, let me know what you think.

### CODE REVIEW

**Comment by finagolfin** [comment]
Rebased, @al45tair, would you run the CI on this and review?

**Comment by shahmishal** [comment]
@swift-ci test

**Comment by finagolfin** [comment]
Rebased, updated to latest NDK 27c, and added some runtime libraries that were being listed over and over again when linking test runners for Android to the list of libraries to be de-duplicated.

@etcwilde, would you review?

**Review by etcwilde — COMMENTED** [review]
I think the changes look fine. I'd like to get approval from @compnerd once he's back though. Is this needed immediately, or can it wait a couple more weeks?

**Comment by finagolfin** [comment]
It can wait.

The BC guys like Saleem haven't been too active on github lately, something going on?

**Comment by finagolfin** [comment]
@hyp, if you're active again, would you review?

**Comment by finagolfin** [comment]
Rebased and enabled another C++ Interop test that now passes on Android.

Ping @etcwilde, it's been three weeks with no response, it would be good to get this in before the branch. If the BC people have any complaint later, which is extremely unlikely, we can always modify the single module map change then.

File: stdlib/public/Platform/android.modulemap [hunk 1]
```diff
@@ -556,6 +556,14 @@ module _bits_sa_family_t [system] {
   header "bits/sa_family_t.h"
   export *
 }
+
+module _bits_sockaddr_storage [system] {
```
Comment (compnerd) [inline stdlib/public/Platform/android.modulemap:560]: Is this backwards compatible?

File: test/Interop/CxxToSwiftToCxx/bridge-cxx-struct-back-to-cxx.swift [hunk 2]
```diff
@@ -21,8 +21,6 @@
 // RUN: %target-interop-build-clangxx -fsyntax-only -x c++-header %t/full-cxx-swift-cxx-bridging.h -std=gnu++17 -c -fmodules -fcxx-modules -I %t 
 // FIXME: test c++20 (rdar://117419434)
 
-// XFAIL: OS=linux-android, OS=linux-androideabi
```
Comment (compnerd) [inline test/Interop/CxxToSwiftToCxx/bridge-cxx-struct-back-to-cxx.swift:24]: Is this backwards compatible?

**Review by compnerd — COMMENTED** [review]
It is unclear if the test/modularization changes would result in issues for r26 or not. In essence, does this prevent the use of r26 or is it simply updating the documentation to reference r27? r26 also is a LTS release, so I'm not sure if we should be upgrading so aggressively. However, updating to also support r27 is preferable.

File: stdlib/public/Platform/android.modulemap (hunk 1 above)
Comment (finagolfin) [reply stdlib/public/Platform/android.modulemap:560]: With NDK 26? As I wrote above, "it still worked fine with NDK 26 too."

File: test/Interop/CxxToSwiftToCxx/bridge-cxx-struct-back-to-cxx.swift (hunk 2 above)
Comment (finagolfin) [reply test/Interop/CxxToSwiftToCxx/bridge-cxx-struct-back-to-cxx.swift:24]: With NDK 26? I think this was fixed by the new Android overlay, so it should be.

**Comment by finagolfin** [comment]
> It is unclear if the test/modularization changes would result in issues for r26 or not.

The modularization change was explicitly tested with NDK 26, as I noted months ago above, "it still worked fine with NDK 26 too." The test change I've only been running with NDK 27 lately, but I believe it works for both.

> In essence, does this prevent the use of r26 or is it simply updating the documentation to reference r27?

Neither, both NDKs should be fine and it is clearly updating more than the docs.

> r26 also is a LTS release, so I'm not sure if we should be upgrading so aggressively

[NDK 27 is the current LTS release](https://developer.android.com/ndk/downloads), which means [NDK 26 is no longer officially supported](https://github.com/android/ndk/wiki/Unsupported-Downloads#r26d). Nevertheless, this pull should work fine with both.

> However, updating to also support r27 is preferable.

Yes, that is what this pull does.

**Comment by finagolfin** [comment]
Ping @compnerd, answered all your questions, would be good to get this in.

**Comment by finagolfin** [comment]
@egorzhdan, need a CI run here.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by hyp** [comment]
I validated this on a windows host with NDK 26 , this looks good to me

**Review by compnerd — APPROVED** [review]

**Review by etcwilde — APPROVED** [review]

**Comment by finagolfin** [comment]
Alright, passed CI, @ktoso, would you merge? I figure everybody else in this thread is in the US and probably sleeping now.

**Review by ktoso — APPROVED** [review]
Ok, happy to since seems others approved already

**Comment by finagolfin** [comment]
Thanks, @etcwilde, @compnerd, @hyp, and @ktoso, wanted to get this in before the upcoming branch.

**Pull request merged**